File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        count = 0
        current = head
        
        while current != None:
            count+=1
            current = current.next
        
        if count == n:
            head = head.next
            return head

        count -= (n+1)

        current = head

        while count>0:
            count-=1
            current = current.next

        hold = current.next

        current.next = None

        if hold.next != None:
            current.next = hold.next

        return head

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: *mut ListNode,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: std::ptr::null_mut(), val }
//     }
// }

impl Solution {
    pub fn has_cycle(mut head: *mut ListNode) -> bool {
        let mut node_set : HashSet<*const ListNode> = HashSet::new();
        
        while !head.is_null() {
            let ptr : *const ListNode = head as *const ListNode;
            if node_set.contains(&ptr) {
                return true;
            }
            node_set.insert(ptr);
            unsafe {
                head = (*head).next;
            }
        }
        return false;
    }
}
